src/mte_intro.adoc

```diff
@@ -1,79 +1,77 @@
-  accessed by it, then it may lead to stack invalid memory access issues.
+* **Stack**: The operating system allocates a region to be used as the stack of
+  the program. When a function is called, a stack frame is created at the top
+  of the stack. This is used to hold local variables, which are accessed
+  relative to the stack pointer, which designates the top of the stack.
+  Upon return from the function, the frame is popped from the stack by
+  adjusting the stack pointer. Within a stack frame, the compiler can reuse
+  the same memory for local variables that have non-overlapping lifetimes
+  (known as stack coloring, implemented e.g. in LLVM StackColoring
+  cite:[LLVM_StackColoring]). The lifetime an object on the stack is
+  governed by the programming language's scoping rules (e.g. C++ scoping rules
+  cite:[CppScope]).
 
-* Globals - Pointers to global memory are valid throughout the runtime of the
-  program and thus invalid memory access to global memory is not possible.
-  Although if a program is accessing global memory in a way that can lead to
-  out of bounds access, that can be leveraged to access address space of the
-  program which otherwise wasn’t possible.
+* **Globals**: Storage for global variables is allocated by the operating system
+  on process startup (or when the dynamic object is loaded). Generally, the
+  storage for globals is destroyed on process shutdown (depending on the OS,
+  `dlclose` might cause it to be destoyed earlier).
+  The lifetime of a global ends when the process is terminated, or the dynamic
+  object is unloaded.
 
-Two prominent patterns of invalid memory accesses are following (but not
-limited to)
+Two prominent patterns of invalid memory accesses are:
 
-* Stale pointer accesses (temporal): Memory was freed (heap) or became out of
-  scope (stack) and due to a bug in software, pointer is still kept around and
-  reachable via some control flow. This can be termed as use after free or out
-  of scope accesses. With respect to heap memory such issues are commonly
-  called `use after free` (UAF). In case of stack, such issues are termed as
-  `stack use after return`, `stack use after scope`.
+* **Temporal**: A pointer is dereferenced past the lifetime of the object
+  it refers to. With respect to heap memory, such issues are commonly
+  called "use after free" (UAF). In case of stack, such issues are termed as
+  "stack use after return", "stack use after scope".
 
-* Incorrect pointer construction (spatial): Due to bug in bounds checking, an
-  incorrect pointer can be constructed and thus leading to invalid memory
-  access. Similarly due to a bug in arithmetics (integer overflow) or bad
-  casting (signed to unsigned), bad offsets can be generated and eventually
-  leading to incorrect pointer construction and out of bounds accesses. More
-  commonly they are termed `stack` or `heap buffer overflows`.
+* **Spatial**: A pointer outside of the bounds of the object it refers to
```

Comment:
The earlier paragraphs about lifetime are largely concerned only about temporal safety. It might be easier to understand if the temporal/spatial dimensions are introduced earlier.